docs/DesignDocs/ExtendedEnclaveInitializationData.md

```diff
@@ -0,0 +1,154 @@
+
+First, before an enclave image with EEID is loaded, we patch the size settings from the signed image with the instance-specific settings.
+This actually alters the computation of the `MRENCLAVE` as the sequence of calls to `EADD/EEXTEND` is modified compared to the values used during signing.
+```C
+    if (eeid)
+    {
+        if (props.header.size_settings.num_heap_pages != 0 ||
+            props.header.size_settings.num_stack_pages != 0 ||
+            props.header.size_settings.num_tcs != 0)
+            OE_RAISE(OE_INVALID_PARAMETER);
+
+        props.header.size_settings.num_heap_pages =
+            eeid->size_settings.num_heap_pages;
+        props.header.size_settings.num_stack_pages =
+            eeid->size_settings.num_stack_pages;
+        props.header.size_settings.num_tcs = eeid->size_settings.num_tcs;
+
+        // patch
+        elf64_sym_t sym_props = {0};
+        const elf64_t* eimg = &oeimage.u.elf.elf;
+        if (elf64_find_symbol_by_name(
+                eimg, "oe_enclave_properties_sgx", &sym_props) == 0)
+        {
+            uint64_t* sym_props_addr = NULL;
+            sym_props_addr =
+                (uint64_t*)(oeimage.image_base + sym_props.st_value);
+            oe_sgx_enclave_properties_t* p =
+                (oe_sgx_enclave_properties_t*)sym_props_addr;
+            p->header.size_settings = props.header.size_settings;
+        }
+    }
+```
+
+Then, after the enclave image is loaded but before it is initialized, we add the additional memory pages containing the contents of the `oe_eeid_t` struct:
+
+```C
+        oe_sha256_context_t* hctx = &context->hash_context;
+        sgx_sigstruct_t* sigstruct = (sgx_sigstruct_t*)properties->sigstruct;
+        memcpy(eeid->sigstruct, (uint8_t*)sigstruct, sizeof(sgx_sigstruct_t));
+        oe_sha256_save(hctx, eeid->hash_state_H, eeid->hash_state_N);
+        eeid->data_vaddr = *vaddr;
+
+        uint64_t ee_sz = sizeof(oe_eeid_t) + eeid->data_size;
+        uint64_t epg_sz = eeid_pages_size(eeid);
+        uint64_t num_pages = epg_sz / OE_PAGE_SIZE;
+        assert(*vaddr == enclave_end - epg_sz);
+
+        oe_page_t* pages = (oe_page_t*)eeid;
+        if (ee_sz < epg_sz)
+        {
+            oe_page_t* tmp = (oe_page_t*)calloc(1, epg_sz);
+            memcpy(tmp, eeid, ee_sz);
+            pages = tmp;
+        }
+
+        OE_CHECK(_add_extra_data_pages(
+            context, enclave->addr, pages, num_pages, vaddr));
+
+        if (ee_sz < epg_sz)
+            free(pages);
+
+        OE_SHA256 ext_mrenclave;
+        oe_sha256_final(hctx, &ext_mrenclave);
+```
+
+Finally, since the changes invalidate the signature of the base image, we need to dynamically re-sign the enclave. For this, we use the `OE_DEBUG_SIGN_KEY`, which is well-known by all OE users. Note that applications should never rely on the apparent `MRSIGNER` of an EEID quote. Indeed, what we reflect to the application when we validate an EEID quote is the `MRSIGNER` of the base image.
```

Comment:
Zero requirement: I don't think so, we just chose to do that in our draft implementation because it was easy to do. Without that, we would have to re-hash the image up to the point before memory pages are added and during verification we would have to extend the image with the original pages that the signature was over.

Enforced: Yes, see `OE_RAISE(OE_INVALID_PARAMETER);` further above.

You're right, for the future we should not rely on those settings being zero, but everything would run a bit faster when they can be.

Reply:
Can we remove this requirement then? Alternatively, have we considered adding a different metadata blob type and explicitly classifying this as a different enclave type?

Reply:
Sure, we can do that. I'd be happy to make it an entirely new enclave type (which could also be sgx-only for now?), although so far I thought of it more of an option to existing types rather than a different type altogether.

Reply:
@CodeMonkeyLeet the idea behind the zero requirement is to ensure that an enclave image meant to be used with EEID is not accidentally loaded without it. In other words, when the image is signed, the signer either decides that the heap & stack configuration is fixed, which makes it impossible to change at load time with EEID, or it set at load time, which then makes the image impossible to load without EEID. Does that make sense?

Now, the idea to sign an invalid configuration is a bit of a hack to implement this concept, and I understand your nervousness. Alternatively, we can add an explicit flag set at enclave signing time that indicates whether the image is meant to be loaded with EEID or not. It is also possible to make this more flexible by having 3 possible settings (EEID disabled, EEID optional, EEID mandatory) but I personally think it is dangerous to make EEID optional when we rely on it for secure key release policies